src/java.base/share/classes/java/io/IO.java

```diff
@@ -0,0 +1,110 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.io;
+
+import jdk.internal.javac.PreviewFeature;
+
+/**
+ * A collection of static convenience methods that provide access to
+ * {@linkplain System#console() system console} for implicitly declared classes.
+ *
+ * <p> Each of this class' methods throws {@link IOError} if the system console
+ * is {@code null}; otherwise, the effect is as if a similarly-named method
+ * had been called on that console.
+ *
+ * <p> Input and output from methods in this class use the character set of
+ * the system console as specified by {@link Console#charset}.
+ *
+ * @since 23
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
+public final class IO {
+
+    private IO() {
+        throw new Error("no instances");
+    }
+
+    /**
+     * Writes a string representation of the specified object to the system
+     * console, terminates the line and then flushes that console.
+     *
+     * <p> The effect is as if {@link Console#println(Object) println(obj)}
+     * had been called on {@code System.console()}.
+     *
+     * @param obj the object to print, may be {@code null}
+     *
+     * @throws IOError if {@code System.console()} returns {@code null},
+     *                 or if an I/O error occurs
+     */
+    public static void println(Object obj) {
+        con().println(obj);
+    }
+
+    /**
+     * Writes a string representation of the specified object to the system
+     * console and then flushes that console.
+     *
+     * <p> The effect is as if {@link Console#print(Object) print(obj)}
+     * had been called on {@code System.console()}.
+     *
+     * @param obj the object to print, may be {@code null}
+     *
+     * @throws IOError if {@code System.console()} returns {@code null},
+     *                 or if an I/O error occurs
+     */
+    public static void print(Object obj) {
+        con().print(obj);
+    }
+
+    /**
+     * Writes a prompt as if by calling {@code print}, then reads a single line
+     * of text from the system console.
+     *
+     * <p> The effect is as if {@link Console#readln(String) readln(prompt)}
+     * had been called on {@code System.console()}.
+     *
+     * @param prompt the prompt string, may be {@code null}
+     *
+     * @return a string containing the line read from the system console, not
+     * including any line-termination characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if {@code System.console()} returns {@code null},
+     *                 or if an I/O error occurs
+     */
+    public static String readln(String prompt) {
```

Comment:
Did we consider Optional<String> here? Maybe that is to complicated for an onramp?

Reply:
Have we considered it? Yes, momentarily. Will we implement it here? No, certainly.

Reply:
The precedent to return `null` from something that reads a line has already been set by `BufferedReader.readLine()` and `Console.readLine()`.

`Optional` is indeed intended for return values but it probably works better for things that are totally new, such as various streams operation (for which `Optional` was introduced), modules, the Classfile API, HTTP client, etc. And to use `Optional` effectively one needs to know how to use lambdas, the teaching of which is probably at odds with the on-ramp story.